arch: cortex-m0: switch to asm!()

bradjc opened the pull request:
### Pull Request Overview

This pull request implements #1923 for cortex-m0.


### Testing Strategy

None? We don't have any cortex-m0 boards.


### TODO or Help Wanted

Help trying to verify my changes are OK.


### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

ppannuto commented on arch/cortex-m0/src/lib.rs:
```diff
@@ -162,47 +162,56 @@ pub unsafe extern "C" fn switch_to_user(
+    ",
+    inout("r0") user_stack,
+    in("r1") process_regs,
+    out("r3") _, out("r4") _, out("r5") _, out("r8") _, out("r9") _,
```
`r7`?

ppannuto commented on arch/cortex-m0/src/lib.rs:
```diff
@@ -286,19 +293,16 @@ _hardfault_msp:
-    );
+    ",
+    out("r0") faulting_stack, out("r1") kernel_stack,
+    out("r2") _, out("r3") _);
```
I believe this can add the `nomem` and `nostack` options

bradjc commented:
Updated. Split hardfault handler to mimic cortex-m version since asm!() macro can be the only thing in a #[naked] function. Also save both r6 and r7 in context switch.

hudson-ayers commented on arch/cortex-m0/src/lib.rs:
```diff
@@ -111,7 +111,7 @@ MEXC_RETURN_MSP:
 MEXC_RETURN_PSP:
   .word 0xFFFFFFFD"
-    : : : : "volatile");
+    );
```
```suggestion
    , options(noreturn)
    );
```
the assembly manually jumps out. Also, naked functions require a single `noreturn` inline asm block -- it is UB for an inline asm block inside a naked function to return.

This suggestion will probably need formatting

hudson-ayers commented on arch/cortex-m0/src/lib.rs:
```diff
@@ -144,7 +144,7 @@ EXC_RETURN_MSP:
   .word 0xFFFFFFFD
   "
-  : : : : "volatile"  );
+    );
```
```suggestion
    , options(noreturn)
    );
```

phil-levis commented:
bors r+
